### ISSUE TITLE
Fix closure marshalling from Dynlink-loaded code on Windows

### ISSUE THREAD

**Issue by dra27** [ocaml/ocaml#10884]
The lib-dynlink-native test is failing on mingw-w64 with `Invalid_argument("output_value: abstract value (outside heap)")`.

The direct cause of the failure is because the new lf_skiplist used in codefrag discards duplicate keys, but the underlying cause was present before. There are three possible things to fix, so looking for guidance on exactly which to do before opening PRs.

Here's roughly what's happening, with sample addresses from my box. `plugin.so` is loaded and when `run_shared_startup` a code fragment is registered starting at 0x31c301510 and also ending at 0x31c301510. When `run` is then executed, a code fragment is registered at the same starting address, but ending at 0x31c301606. This is not new - 4.14 (i.e. before #10831) does the same thing.

However, the new lockfree skiplist discards duplicate entries to a skiplist where the previous implementation instead updates the data.

There are three possible fixes, but I'm not sure which ones are wanted:

## 1. Tweak codefrag to be a little more liberal
Have `pc <= cf->code_end` instead of:
https://github.com/ocaml/ocaml/blob/1546e1f86f87485637719c86d92f32df9292997f/runtime/codefrag.c#L105
and/or have `caml_register_code_fragment` reject fragments with the same start and end address? Note that tweaking codefrag is not enough to fix this (it's just whether those functions _technically_ ought to be more liberal, or be doing more checking).
## 2. Make the `caml_lf_skiplist_insert` like `caml_skiplist_insert` for duplicates
```diff
--- a/runtime/lf_skiplist.c
+++ b/runtime/lf_skiplist.c
@@ -353,8 +353,9 @@ int caml_lf_skiplist_insert(struct lf_skiplist *sk, uintnat key, uintnat data) {
     struct lf_skipcell *succ;

     if (found) {
-      /* node already exists, we don't need to do anything */
-      return 0;
+      /* Already present; update data */
+      atomic_store_explicit(&succs[0]->data, data, memory_order_release);
+      return 1;
     } else {
       /* node does not exist. We need to generate a random top_level and
        * construct a new node. The new node's forward array (which contains the
```
Two racing insertions may both update the data; only one wins. For codefrag, the race is impossible™.
## 3. Stop registering two code fragments at the same start address

I don't follow why the same address ends up being registered twice. If we adopt a fix loosely found from #4690:

```diff
diff --git a/asmcomp/amd64/emit.mlp b/asmcomp/amd64/emit.mlp
index ef768b11d..be216ac44 100644
--- a/asmcomp/amd64/emit.mlp
+++ b/asmcomp/amd64/emit.mlp
@@ -1048,7 +1048,7 @@ let end_assembly() =
   end;

   emit_named_text_section (Compilenv.make_symbol (Some "code_end"));
-  if system = S_macosx then I.nop ();
+  if system = S_macosx || system = S_mingw64 then I.nop ();
   (* suppress "ld warning: atom sorting error" *)

   emit_global_label "code_end";
```

then the first code fragment is 0x24cb61510 ending at 0x24cB61511 and the next code fragment starts at 0x24cb61520 ending at 0x24cb61756 and so everything works. However, that fix is an unexpected sledgehammer. Another possibility might be:

```diff
--- a/runtime/dynlink_nat.c
+++ b/runtime/dynlink_nat.c
@@ -117,7 +117,7 @@ CAMLprim value caml_natdynlink_run(value handle_v, value symbol) {

   sym = optsym("__code_begin");
   sym2 = optsym("__code_end");
-  if (NULL != sym && NULL != sym2) {
+  if (NULL != sym && NULL != sym2 && sym != sym2) {
     caml_register_code_fragment((char *) sym, (char *) sym2,
                                 DIGEST_LATER, NULL);
   }
```

**Comment by sadiqj** [comment]
Regardless of our choice for fixing codefrag, I think it is worth doing option 2. It is an inconsistency between the  sequential and lock-free skiplists that might trip up future users.

**Comment by xavierleroy** [comment]
I think we should not register empty code fragments, since there will never be a code pointer inside such a fragment.  So I support the "another possibility" at the end of approach 3.  I think this is the only case where we can have two code fragments at the same starting address.

**Comment by dra27** [comment]
OK - PRs to follow shortly!

**Comment by xavierleroy** [comment]
Looks like this issue was addressed in #10902 and  in #10903.
